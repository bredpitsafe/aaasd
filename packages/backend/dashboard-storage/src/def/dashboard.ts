import type {
    TKind,
    TStatus,
} from '@grpc-schemas/dashboard_storage-api-sdk/index.services.dashboard_storage.v1';

import type { Kind, Status } from './response';

export type TDbDashboard = {
    id: string;
    legacyId?: number;
    name: string;
    kind: Kind;
    status: Status;
    config: string;
    digest: string;
    // This is update time actually
    insertionTime: string;
};

export type TGrpcDashboard = Omit<TDbDashboard, 'kind' | 'status'> & {
    kind: TKind;
    status: TStatus;
};
