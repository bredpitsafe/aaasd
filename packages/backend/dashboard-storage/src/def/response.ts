/* eslint-disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

/**
 * @deprecated
 */
export type DashboardItem = DashboardsListItem & {
  draftDigest: string;
};


// TODO: remove this file after enums transfer into more appropriate place and tests fix
/**
 * @deprecated
 */
export interface Response {
  error?: ErrorResponse;
  payload?:
    | UnsubscribeResponse
    | PingResponse
    | HeartbeatResponse
    | AuthenticateResponse
    | LogoutResponse
    | SubscribeToDashboardResponse
    | SubscribeToDashboardsListResponse
    | FetchDashboardConfigResponse
    | FetchDashboardDraftResponse
    | CreateDashboardResponse
    | ImportDashboardResponse
    | UpdateDashboardResponse
    | UpdateDashboardDraftResponse
    | ResetDashboardDraftResponse
    | UpdateDashboardPermissionsResponse
    | UpdateDashboardShareSettingsResponse
    | DeleteDashboardResponse
    | SubscribeToDashboardPermissionsResponse
    | FetchDashboardIdByLegacyIdResponse
    | RenameDashboardResponse
    | UpdateDashboardScopeBindingResponse;
  timestamp: string;
  traceId: string;
  correlationId: number;
  state: ResponseState;
}
/**
 * @deprecated
 */
export interface ErrorResponse {
  kind: string;
  description: string;
  args?: {
    [k: string]: unknown;
  };
}
/**
 * @deprecated
 */
export interface UnsubscribeResponse {
  type: "Unsubscribed";
}
/**
 * @deprecated
 */
export interface PingResponse {
  type: "Pong";
}
/**
 * @deprecated
 */
export interface HeartbeatResponse {
  type: "Heartbeat";
}
/**
 * @deprecated
 */
export interface AuthenticateResponse {
  type: "Authenticated";
  username: string;
}
/**
 * @deprecated
 */
export interface LogoutResponse {
  type: "LoggedOut";
}
/**
 * @deprecated
 */
export interface SubscribeToDashboardResponse {
  type: "Dashboard";
  dashboard: DashboardItem;
}
/**
 * @deprecated
 */
export interface DashboardsListItem {
  id: string;
  name: string;
  kind: Kind;
  status: Status;
  digest?: string;
    // This is update time actually
  insertionTime: string;
  hasDraft: boolean;
  sharePermission: SharePermission;
  permission: Permission;
  owners: User[];
  permissionsCount?: number;
  scopes: Scope[];
}
/**
 * @deprecated
 */
export interface User {
  user: string;
}
/**
 * @deprecated
 */
export interface Scope {
  [k: string]: string | number | boolean | null;
}
/**
 * @deprecated
 */
export interface SubscribeToDashboardsListResponse {
  type: "DashboardsList";
  list: DashboardsListItem[];
}
/**
 * @deprecated
 */
export interface FetchDashboardConfigResponse {
  type: "DashboardConfig";
  config: string;
  digest: string;
}
/**
 * @deprecated
 */
export interface FetchDashboardDraftResponse {
  type: "DashboardDraft";
  draft: string;
  digest: string;
}
/**
 * @deprecated
 */
export interface CreateDashboardResponse {
  type: "DashboardCreated";
  id: string;
}
/**
 * @deprecated
 */
export interface ImportDashboardResponse {
  type: "DashboardImported";
  id: string;
}
/**
 * @deprecated
 */
export interface UpdateDashboardResponse {
  type: "DashboardUpdated";
}
/**
 * @deprecated
 */
export interface UpdateDashboardDraftResponse {
  type: "DashboardDraftUpdated";
}
/**
 * @deprecated
 */
export interface ResetDashboardDraftResponse {
  type: "DashboardDraftReset";
}
/**
 * @deprecated
 */
export interface UpdateDashboardPermissionsResponse {
  type: "PermissionsUpdated";
}
/**
 * @deprecated
 */
export interface UpdateDashboardShareSettingsResponse {
  type: "DashboardShareSettingsUpdated";
}
/**
 * @deprecated
 */
export interface DeleteDashboardResponse {
  type: "DashboardDeleted";
}
/**
 * @deprecated
 */
export interface SubscribeToDashboardPermissionsResponse {
  type: "DashboardPermissionsList";
  list: {
    user: string;
    permission: string;
  }[];
}
/**
 * @deprecated
 */
export interface FetchDashboardIdByLegacyIdResponse {
  type: "DashboardId";
  id: string;
}
/**
 * @deprecated
 */
export interface RenameDashboardResponse {
  type: "DashboardRenamed";
}
/**
 * @deprecated
 */
export interface UpdateDashboardScopeBindingResponse {
  type: "DashboardScopeBindingUpdated";
  platformTime: string;
}

// TODO: move this enums to the approriate place
export const enum Kind {
  User = "User",
  Robot = "Robot"
}
export const enum Status {
  Active = "Active",
  Suspended = "Suspended",
  Archived = "Archived"
}
export const enum SharePermission {
  None = "None",
  Viewer = "Viewer",
  Editor = "Editor"
}
export const enum Permission {
  None = "None",
  Viewer = "Viewer",
  Editor = "Editor",
  Owner = "Owner"
}
/**
 * @deprecated
 */
export const enum ResponseState {
  Done = "Done",
  InProgress = "InProgress"
}
