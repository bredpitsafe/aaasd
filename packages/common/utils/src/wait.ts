import { firstValueFrom, interval, map } from 'rxjs';

export const wait = (ms = 100): Promise<void> =>
    firstValueFrom<void>(interval(ms).pipe(map(() => undefined)));
