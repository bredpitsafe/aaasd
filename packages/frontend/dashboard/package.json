{
  "name": "@frontend/dashboard",
  "version": "1.0.1",
  "description": "dashboard page",
  "scripts": {
    "build": "NODE_ENV=production npx webpack",
    "jsonschema": "./src/schemas/generateSchemaValidators.sh",
    "lint": "eslint './src/**/*.{js,ts,tsx}' --quiet",
    "server": "npx webpack serve",
    "test_unit": "jest --passWithNoTests",
    "types": "tsc --noEmit"
  },
  "dependencies": {
    "@backend/bff": "^0.0.1",
    "@common/rx": "^1.0.0",
    "@common/types": "^1.0.0",
    "@common/utils": "^1.0.0",
    "@frontend/charter": "^1.0.1",
    "@frontend/common": "^1.0.1",
    "@frontend/e2e-tests": "^1.0.1",
    "@frontend/ag-grid": "^1.0.0",
    "idb": "^7.0.0"
  }
}
