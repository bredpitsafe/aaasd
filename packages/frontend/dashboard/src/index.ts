import '@frontend/common/src/utils/Rx/internalProviders';

import { createLocalWorkerActor } from '@frontend/common/src/actors/LocalWorker';

import { createActorDashboardsTab } from './actor';
import { createActorFullDashboards } from './actors/FullDashboards';

const workerRoot = createLocalWorkerActor();
const actorTab = createActorDashboardsTab();
const fullDashboardActor = createActorFullDashboards();

/*connectActorToActor(fullDashboardActor, actorTab);

connectActorToActor(workerRoot, {
    transmitter: actorTab,
    map(envelope) {
        return FullDashboardsDescriptorNames.has(envelope.type) ? undefined : envelope;
    },
});
connectActorToActor(workerRoot, fullDashboardActor);*/

workerRoot.launch();
fullDashboardActor.launch();
actorTab.launch();
