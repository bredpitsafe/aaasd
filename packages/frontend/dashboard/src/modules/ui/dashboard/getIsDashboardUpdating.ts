import { constantNormalizer } from '@frontend/common/src/utils/observable/memo.ts';
import { createRemoteProcedureCall } from '@frontend/common/src/utils/RPC/createRemoteProcedureCall.ts';

import { SubscribeToDashboardUpdateProgressProcedureDescriptor } from '../../../actors/FullDashboards/descriptors.ts';

export const ModuleSubscribeToDashboardsUpdateProgress = createRemoteProcedureCall(
    SubscribeToDashboardUpdateProgressProcedureDescriptor,
)({
    dedobs: {
        normalize: constantNormalizer,
        resetDelay: 0,
        removeDelay: 0,
    },
});
