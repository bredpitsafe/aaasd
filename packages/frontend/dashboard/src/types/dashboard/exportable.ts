import type { TExportableGridSettings } from '../layout/exportable';
import type { TExportablePanel } from '../panel/exportable';

export type TExportableDashboardEditorWrapped = {
    dashboard: TExportableDashboardEditor;
};

export type TExportableDashboardWrapped = {
    dashboard: TExportableDashboard;
};

export type TExportableDashboardFileWrapped = {
    dashboard: TExportableDashboardFile;
};

type TExportableDashboardEditor = {
    version?: string;
    activeLayout?: string;
    name: string;
    grid: TExportableGridSettings;
    panels: {
        panel: (Omit<TExportablePanel, 'panelId'> & { comment: string })[];
    };
};

type TExportableDashboardFile = {
    version?: string;
    activeLayout?: string;
    name: string;
    grid: TExportableGridSettings;
    panels: {
        panel: Omit<TExportablePanel, 'panelId'>[];
    };
};

/** @public */
export type TExportableDashboard = {
    version?: string;
    activeLayout?: string;
    name: string;
    grid: TExportableGridSettings;
    panels: {
        panel: TExportablePanel[];
    };
};
