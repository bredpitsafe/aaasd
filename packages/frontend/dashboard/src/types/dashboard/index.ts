import type { TStorageDashboardName } from '@frontend/common/src/types/domain/dashboardsStorage';
import type { TSemverVersion } from '@frontend/common/src/utils/Semver/def';

import type { TGridSettings } from '../layout';
import type { TPanel } from '../panel';

export type TDashboard = {
    version?: TSemverVersion;
    activeLayout: undefined | string;
    name: TStorageDashboardName;
    grid: TGridSettings;
    panels: TPanel[];
};
