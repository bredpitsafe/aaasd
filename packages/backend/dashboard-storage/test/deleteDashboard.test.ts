import { wait } from '@common/utils/src/wait.js';
import { afterAll, describe, expect, it, vi } from 'vitest';

import { SharePermission } from '../src/def/response.js';
import { Client, users } from './client/index.js';
import dashboard from './fixtures/dashboard.json';

describe('DeleteDashboard', () => {
    const client: Client = new Client({ authenticate: true });
    const otherClient: Client = new Client({ authenticate: true, as: users.user2.username });

    it('deletes own dashboard', async () => {
        const { id } = (await client.request(dashboard)).payload;
        expect(id).not.toBeUndefined();

        const list$ = client.send({ type: 'SubscribeToDashboardsList' });
        const subscriber = vi.fn();
        const subscription = list$.subscribe(subscriber);

        const response = await client.request({ type: 'DeleteDashboard', id });
        expect(response.state).toBe('Done');
        expect(response.payload.type).toBe('DashboardDeleted');

        await wait();
        expect(subscriber.mock.calls).toHaveLength(2);
        const firstResponse = subscriber.mock.calls[0][0];
        expect(
            firstResponse.payload.list.find((item: { id: string }) => item.id === id),
        ).not.toBeUndefined();

        const secondResponse = subscriber.mock.calls[1][0];
        expect(
            secondResponse.payload.list.find((item: { id: string }) => item.id === id),
        ).toBeUndefined();

        subscription.unsubscribe();
    });

    it('deletes dashboard from list after sharing', async () => {
        const { id } = (await client.request(dashboard)).payload;
        const list$ = client.send({ type: 'SubscribeToDashboardsList' });
        const subscriber = vi.fn();
        const subscription = list$.subscribe(subscriber);

        const response = await client.request({ type: 'DeleteDashboard', id });
        expect(response.state).toBe('Done');

        await wait();
        expect(subscriber.mock.calls).toHaveLength(2);
        const firstResponse = subscriber.mock.calls[0][0];
        const secondResponse = subscriber.mock.calls[1][0];
        expect(
            firstResponse.payload.list.find((item: { id: string }) => item.id === id),
        ).toMatchObject({
            id,
        });
        expect(
            secondResponse.payload.list.find((item: { id: string }) => item.id === id),
        ).toBeUndefined();

        subscription.unsubscribe();
    });

    it('fails to delete other user`s dashboard', async () => {
        const { id } = (await client.request(dashboard)).payload;
        expect(id).not.toBeUndefined();

        const response = await otherClient.request({ type: 'DeleteDashboard', id });

        expect(response.error).toMatchSnapshot();
    });

    it('fails to delete already deleted dashboard', async () => {
        const { id } = (await client.request(dashboard)).payload;
        expect(id).not.toBeUndefined();

        const response = await client.request({ type: 'DeleteDashboard', id });
        expect(response.state).toBe('Done');

        const secondResponse = await client.request({ type: 'DeleteDashboard', id });
        expect(secondResponse.error).toMatchSnapshot();
    });

    // This test case doesn't work for now (as expected).
    it.skip('deletes shared dashboard from anyone', async () => {
        // Create dashboard
        const { id } = (await client.request(dashboard)).payload;

        // Share dashboard to other users
        await client.request({
            type: 'UpdateDashboardShareSettings',
            id,
            sharePermission: SharePermission.Viewer,
        });

        // Subscribe to dashboard as other user
        const dashboard$ = otherClient.send({
            type: 'SubscribeToDashboard',
            id,
        });
        const subscriber = vi.fn();
        const subscription = dashboard$.subscribe(subscriber);

        await wait();

        // Unshare dashboard
        await client.request({
            type: 'DeleteDashboard',
            id,
        });

        await wait();
        // Check that subscription has ended with error
        expect(subscriber.mock.calls).toHaveLength(2);
        const firstResponse = subscriber.mock.calls[0][0];
        const secondResponse = subscriber.mock.calls[1][0];
        expect(firstResponse.state).toBe('InProgress');
        expect(secondResponse.state).toBe('Done');
        expect(secondResponse.error).toMatchSnapshot();

        subscription.unsubscribe();
    });

    afterAll(() => {
        client.close();
        otherClient.close();
    });
});
