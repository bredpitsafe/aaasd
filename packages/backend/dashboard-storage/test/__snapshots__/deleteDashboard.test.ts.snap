// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DeleteDashboard > fails to delete already deleted dashboard 1`] = `
{
  "description": "Dashboard delete failed: The dashboard doesnt exist or has been deleted before.",
  "kind": "Authorization",
}
`;

exports[`DeleteDashboard > fails to delete other user\`s dashboard 1`] = `
{
  "description": "Dashboard delete failed: The dashboard doesnt exist or has been deleted before.",
  "kind": "Authorization",
}
`;
